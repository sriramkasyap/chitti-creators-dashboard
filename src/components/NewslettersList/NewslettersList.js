import { useContext, useEffect, useMemo } from "react";
import { Flex, Text, Image, Link as ChakraLink } from "@chakra-ui/react";
import Link from "next/link";

import Table from "../common/Table/Table";
import Pagination from "../common/Pagination/Pagination";

import { getFormattedDate, showNotification } from "../../utils";
import { NewslettersContext } from "../../../contexts/NewslettersContext";

const NewslettersList = () => {
  const {
    newsletters,
    isLoading,
    error,
    getNewsletterLink,
    pagination,
    setPagination,
    totalCount,
  } = useContext(NewslettersContext);

  useEffect(() => {
    if (error) {
      showNotification(error);
    }
  }, [error]);

  const data = useMemo(() => {
    return newsletters.map((newsletter) => ({
      newsletterName: newsletter.reference,
      newsletterSubject: newsletter.emailSubject,
      createdDate: getFormattedDate(newsletter.createdAt),
      status: newsletter.status.toUpperCase(),
      isEditable:
        newsletter.status === "draft" ? (
          getNewsletterLink(newsletter._id)
        ) : (
          <Text>&mdash;</Text>
        ),
    }));
  }, [newsletters]);

  const columns = useMemo(
    () => [
      {
        Header: "Newsletter",
        accessor: "newsletterName",
      },
      {
        Header: "Subject",
        accessor: "newsletterSubject",
      },
      {
        Header: "Created At",
        accessor: "createdDate",
      },
      {
        Header: "Status",
        accessor: "status",
      },

      {
        Header: "Edit Newsletter",
        accessor: "isEditable",
      },
    ],
    []
  );

  return (
    <Flex flexDir="column" w={["", "100%"]}>
      {newsletters && newsletters.length > 0 ? (
        <Flex
          flexDir="column"
          justifyContent={isLoading ? "center" : "flex-start"}
          alignItems="center"
          h="auto"
          w="100%"
        >
          {isLoading ? (
            <Image
              data-testid="loading-image"
              src="loader_black.gif"
              h="5rem"
            />
          ) : (
            <Table
              columns={columns}
              data={data}
              size={("sm", "sm", "md", "lg")}
            />
          )}

          {totalCount > newsletters.length && (
            <Pagination
              {...pagination}
              totalCount={totalCount}
              setPagination={setPagination}
            />
          )}
        </Flex>
      ) : (
        <>
          {!isLoading && (
            <Text data-testid="no-newsletters-data" color="bright.gray">
              You haven&apos;t created any newsletters yet.{" "}
              <Link href="/newsletters/new" as="/newsletters/new">
                <ChakraLink textDecor="underline">Create One now.</ChakraLink>
              </Link>
            </Text>
          )}
        </>
      )}
    </Flex>
  );
};

export default NewslettersList;
