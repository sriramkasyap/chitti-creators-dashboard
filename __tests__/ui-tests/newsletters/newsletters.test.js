import { render, cleanup } from "@testing-library/react";
import "@testing-library/jest-dom/extend-expect";

import NewslettersList from "../../../src/components/NewslettersList/NewslettersList";
import { NewslettersContext } from "../../../contexts/NewslettersContext";

const newsletters = [
  {
    reference: "Newsletter 1",
    emailSubject: "Newsletter 1",
    createdAt: "2021-07-12T09:54:52.918Z",
    status: "draft",
  },
  {
    reference: "Newsletter 2",
    emailSubject: "Newsletter 2",
    createdAt: "2021-07-12T10:23:52.918Z",
    status: "published",
  },
  {
    reference: "Newsletter 3",
    emailSubject: "Newsletter 3",
    createdAt: "2021-07-12T10:15:52.918Z",
    status: "draft",
  },
];

let container;

afterEach(cleanup);

describe("Test Newsletters initial render", () => {
  it("should match the snapshot", () => {
    const contextValue = {
      newsletters,
      isLoading: false,
      error: "",
      getNewsletterLink: jest.fn(),
      pagination: {
        limit: 10,
        page: 0,
      },
      setPagination: jest.fn(),
      totalCount: 40,
    };

    container = render(
      <NewslettersContext.Provider value={contextValue}>
        <NewslettersList />
      </NewslettersContext.Provider>
    );

    expect(container).toMatchSnapshot();
  });
});

describe("Test if no data received", () => {
  it("should render no data text if newsletters is empty", () => {
    const contextValue = {
      newsletters: [],
      isLoading: false,
      error: "",
      getNewsletterLink: jest.fn(),
      pagination: {
        limit: 10,
        page: 0,
      },
      setPagination: jest.fn(),
      totalCount: 0,
    };

    const { getByTestId } = render(
      <NewslettersContext.Provider value={contextValue}>
        <NewslettersList />
      </NewslettersContext.Provider>
    );

    expect(getByTestId("no-newsletters-data")).toBeInTheDocument();
  });

  it("should render loading image on isLoading prop change", () => {
    const contextValue = {
      newsletters,
      isLoading: true,
      error: "",
      getNewsletterLink: jest.fn(),
      pagination: {
        limit: 10,
        page: 0,
      },
      setPagination: jest.fn(),
      totalCount: 0,
    };

    const { getByTestId } = render(
      <NewslettersContext.Provider value={contextValue}>
        <NewslettersList />
      </NewslettersContext.Provider>
    );

    expect(getByTestId("loading-image")).toBeInTheDocument();
  });

  it("should render notification on error prop change", () => {
    const contextValue = {
      newsletters,
      isLoading: false,
      error: "some error",
      getNewsletterLink: jest.fn(),
      pagination: {
        limit: 10,
        page: 0,
      },
      setPagination: jest.fn(),
      totalCount: 0,
    };

    const { getByTestId } = render(
      <NewslettersContext.Provider value={contextValue}>
        <NewslettersList />
      </NewslettersContext.Provider>
    );

    expect(getByTestId("notification")).toBeInTheDocument();
  });
});
