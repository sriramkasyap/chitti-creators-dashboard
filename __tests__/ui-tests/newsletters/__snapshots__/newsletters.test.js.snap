// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Test Newsletters initial render should match the snapshot 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": <body>
    <div
      id="chakra-toast-portal"
    >
      <ul
        id="chakra-toast-manager-top"
        style="position: fixed; z-index: 5500; pointer-events: none; display: flex; flex-direction: column; margin: 0px auto;"
      />
      <ul
        id="chakra-toast-manager-top-left"
        style="position: fixed; z-index: 5500; pointer-events: none; display: flex; flex-direction: column;"
      />
      <ul
        id="chakra-toast-manager-top-right"
        style="position: fixed; z-index: 5500; pointer-events: none; display: flex; flex-direction: column;"
      />
      <ul
        id="chakra-toast-manager-bottom-left"
        style="position: fixed; z-index: 5500; pointer-events: none; display: flex; flex-direction: column;"
      />
      <ul
        id="chakra-toast-manager-bottom"
        style="position: fixed; z-index: 5500; pointer-events: none; display: flex; flex-direction: column; margin: 0px auto;"
      />
      <ul
        id="chakra-toast-manager-bottom-right"
        style="position: fixed; z-index: 5500; pointer-events: none; display: flex; flex-direction: column;"
      />
    </div>
    <div>
      <div
        class="css-h53f3m"
      >
        <div
          class="css-1re57g1"
        >
          <table
            class="chakra-table css-0"
            role="table"
          >
            <thead
              class="css-0"
            >
              <tr
                class="css-0"
                role="row"
              >
                <th
                  class="css-mvn64f"
                  colspan="1"
                  role="columnheader"
                  style="cursor: pointer;"
                  title="Toggle SortBy"
                >
                  <div
                    class="css-u4p24i"
                  >
                    <p
                      class="chakra-text css-0"
                    >
                      Newsletter
                    </p>
                    <span
                      class="css-rkk3mp"
                    />
                  </div>
                </th>
                <th
                  class="css-mvn64f"
                  colspan="1"
                  role="columnheader"
                  style="cursor: pointer;"
                  title="Toggle SortBy"
                >
                  <div
                    class="css-u4p24i"
                  >
                    <p
                      class="chakra-text css-0"
                    >
                      Subject
                    </p>
                    <span
                      class="css-rkk3mp"
                    />
                  </div>
                </th>
                <th
                  class="css-mvn64f"
                  colspan="1"
                  role="columnheader"
                  style="cursor: pointer;"
                  title="Toggle SortBy"
                >
                  <div
                    class="css-u4p24i"
                  >
                    <p
                      class="chakra-text css-0"
                    >
                      Created At
                    </p>
                    <span
                      class="css-rkk3mp"
                    />
                  </div>
                </th>
                <th
                  class="css-mvn64f"
                  colspan="1"
                  role="columnheader"
                  style="cursor: pointer;"
                  title="Toggle SortBy"
                >
                  <div
                    class="css-u4p24i"
                  >
                    <p
                      class="chakra-text css-0"
                    >
                      Status
                    </p>
                    <span
                      class="css-rkk3mp"
                    />
                  </div>
                </th>
                <th
                  class="css-mvn64f"
                  colspan="1"
                  role="columnheader"
                  style="cursor: pointer;"
                  title="Toggle SortBy"
                >
                  <div
                    class="css-u4p24i"
                  >
                    <p
                      class="chakra-text css-0"
                    >
                      Edit Newsletter
                    </p>
                    <span
                      class="css-rkk3mp"
                    />
                  </div>
                </th>
              </tr>
            </thead>
            <tbody
              class="css-0"
              role="rowgroup"
            >
              <tr
                class="css-0"
                role="row"
              >
                <td
                  class="css-j251j9"
                  role="cell"
                >
                  Newsletter 1
                </td>
                <td
                  class="css-j251j9"
                  role="cell"
                >
                  Newsletter 1
                </td>
                <td
                  class="css-j251j9"
                  role="cell"
                >
                  12-07-2021 3:24:52 pm
                </td>
                <td
                  class="css-j251j9"
                  role="cell"
                >
                  DRAFT
                </td>
                <td
                  class="css-j251j9"
                  role="cell"
                >
                  
                </td>
              </tr>
              <tr
                class="css-0"
                role="row"
              >
                <td
                  class="css-j251j9"
                  role="cell"
                >
                  Newsletter 2
                </td>
                <td
                  class="css-j251j9"
                  role="cell"
                >
                  Newsletter 2
                </td>
                <td
                  class="css-j251j9"
                  role="cell"
                >
                  12-07-2021 3:53:52 pm
                </td>
                <td
                  class="css-j251j9"
                  role="cell"
                >
                  PUBLISHED
                </td>
                <td
                  class="css-j251j9"
                  role="cell"
                >
                  <p
                    class="chakra-text css-0"
                  >
                    â€”
                  </p>
                </td>
              </tr>
              <tr
                class="css-0"
                role="row"
              >
                <td
                  class="css-j251j9"
                  role="cell"
                >
                  Newsletter 3
                </td>
                <td
                  class="css-j251j9"
                  role="cell"
                >
                  Newsletter 3
                </td>
                <td
                  class="css-j251j9"
                  role="cell"
                >
                  12-07-2021 3:45:52 pm
                </td>
                <td
                  class="css-j251j9"
                  role="cell"
                >
                  DRAFT
                </td>
                <td
                  class="css-j251j9"
                  role="cell"
                >
                  
                </td>
              </tr>
            </tbody>
          </table>
          <div
            class="css-3mgjd9"
          >
            <button
              class="chakra-button css-1y29bcw"
              data-testid="prev-btn"
              disabled=""
              type="button"
            >
              &lt; Previous
            </button>
            <p
              class="chakra-text css-0"
            >
              Page 
              1
               of 
              4
            </p>
            <button
              class="chakra-button css-1y29bcw"
              data-testid="next-btn"
              type="button"
            >
              Next &gt;
            </button>
          </div>
        </div>
      </div>
    </div>
  </body>,
  "container": <div>
    <div
      class="css-h53f3m"
    >
      <div
        class="css-1re57g1"
      >
        <table
          class="chakra-table css-0"
          role="table"
        >
          <thead
            class="css-0"
          >
            <tr
              class="css-0"
              role="row"
            >
              <th
                class="css-mvn64f"
                colspan="1"
                role="columnheader"
                style="cursor: pointer;"
                title="Toggle SortBy"
              >
                <div
                  class="css-u4p24i"
                >
                  <p
                    class="chakra-text css-0"
                  >
                    Newsletter
                  </p>
                  <span
                    class="css-rkk3mp"
                  />
                </div>
              </th>
              <th
                class="css-mvn64f"
                colspan="1"
                role="columnheader"
                style="cursor: pointer;"
                title="Toggle SortBy"
              >
                <div
                  class="css-u4p24i"
                >
                  <p
                    class="chakra-text css-0"
                  >
                    Subject
                  </p>
                  <span
                    class="css-rkk3mp"
                  />
                </div>
              </th>
              <th
                class="css-mvn64f"
                colspan="1"
                role="columnheader"
                style="cursor: pointer;"
                title="Toggle SortBy"
              >
                <div
                  class="css-u4p24i"
                >
                  <p
                    class="chakra-text css-0"
                  >
                    Created At
                  </p>
                  <span
                    class="css-rkk3mp"
                  />
                </div>
              </th>
              <th
                class="css-mvn64f"
                colspan="1"
                role="columnheader"
                style="cursor: pointer;"
                title="Toggle SortBy"
              >
                <div
                  class="css-u4p24i"
                >
                  <p
                    class="chakra-text css-0"
                  >
                    Status
                  </p>
                  <span
                    class="css-rkk3mp"
                  />
                </div>
              </th>
              <th
                class="css-mvn64f"
                colspan="1"
                role="columnheader"
                style="cursor: pointer;"
                title="Toggle SortBy"
              >
                <div
                  class="css-u4p24i"
                >
                  <p
                    class="chakra-text css-0"
                  >
                    Edit Newsletter
                  </p>
                  <span
                    class="css-rkk3mp"
                  />
                </div>
              </th>
            </tr>
          </thead>
          <tbody
            class="css-0"
            role="rowgroup"
          >
            <tr
              class="css-0"
              role="row"
            >
              <td
                class="css-j251j9"
                role="cell"
              >
                Newsletter 1
              </td>
              <td
                class="css-j251j9"
                role="cell"
              >
                Newsletter 1
              </td>
              <td
                class="css-j251j9"
                role="cell"
              >
                12-07-2021 3:24:52 pm
              </td>
              <td
                class="css-j251j9"
                role="cell"
              >
                DRAFT
              </td>
              <td
                class="css-j251j9"
                role="cell"
              >
                
              </td>
            </tr>
            <tr
              class="css-0"
              role="row"
            >
              <td
                class="css-j251j9"
                role="cell"
              >
                Newsletter 2
              </td>
              <td
                class="css-j251j9"
                role="cell"
              >
                Newsletter 2
              </td>
              <td
                class="css-j251j9"
                role="cell"
              >
                12-07-2021 3:53:52 pm
              </td>
              <td
                class="css-j251j9"
                role="cell"
              >
                PUBLISHED
              </td>
              <td
                class="css-j251j9"
                role="cell"
              >
                <p
                  class="chakra-text css-0"
                >
                  â€”
                </p>
              </td>
            </tr>
            <tr
              class="css-0"
              role="row"
            >
              <td
                class="css-j251j9"
                role="cell"
              >
                Newsletter 3
              </td>
              <td
                class="css-j251j9"
                role="cell"
              >
                Newsletter 3
              </td>
              <td
                class="css-j251j9"
                role="cell"
              >
                12-07-2021 3:45:52 pm
              </td>
              <td
                class="css-j251j9"
                role="cell"
              >
                DRAFT
              </td>
              <td
                class="css-j251j9"
                role="cell"
              >
                
              </td>
            </tr>
          </tbody>
        </table>
        <div
          class="css-3mgjd9"
        >
          <button
            class="chakra-button css-1y29bcw"
            data-testid="prev-btn"
            disabled=""
            type="button"
          >
            &lt; Previous
          </button>
          <p
            class="chakra-text css-0"
          >
            Page 
            1
             of 
            4
          </p>
          <button
            class="chakra-button css-1y29bcw"
            data-testid="next-btn"
            type="button"
          >
            Next &gt;
          </button>
        </div>
      </div>
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
